Enable JIT cage support in Wasm JS String builtins implementation

ast-hugger opened the pull request:
#### 12421d94ba14a8c1b79f348cc29d07bef4f99f03
<pre>
Enable JIT cage support in Wasm JS String builtins implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=297985">https://bugs.webkit.org/show_bug.cgi?id=297985</a>
<a href="https://rdar.apple.com/159302038">rdar://159302038</a>

Reviewed by Keith Miller.

This patch changes the behavior of WasmBuiltinCallee so that if JIT cage
is enabled, instead of using &quot;static&quot; builtin trampolines defined in
InPlaceInterpreter.asm, we JIT-generate their equivalents. This
ensures that the trampolines are callable from Wasm code. A Wasm builtin
entry point is registered as a JIT operation to make it accessible from
the generated trampoline.

The patch also includes refactorings in WebAssemblyBuiltin.{h, cpp} to:
    - Use more appropriate types for the various code pointers held by
      WebAssemblyBuiltin.
    - Clean up hacky pointer tagging in initializeBuiltinImport()
      in WebAssemblyModuleRecord.cpp.
    - Use more descriptive names for the naming macros (the bulk of
      changes in WebAssemblyBuiltin.cpp).

Canonical link: <a href="https://commits.webkit.org/299341@main">https://commits.webkit.org/299341@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3cf633e47cc3c7c19bafeabdf6b6d390318a87d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118652 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38333 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28984 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124832 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70712 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/169dfcbe-3e1f-4011-8cac-b3c435fe0767) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39029 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46915 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90055 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8190c1a0-f359-47e5-9a7c-e8aacb6f727e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121605 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31094 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106386 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70560 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3c554509-f8b1-40bf-baa2-9c0b8c5b1373) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30153 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24497 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68489 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110770 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100532 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24687 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127890 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117166 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45559 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34382 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98696 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45923 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102606 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98477 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25039 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43920 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21922 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42066 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45429 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51107 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145862 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44893 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37503 "Found 1 new JSC binary failure: testapi, Found 18587 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/nativearray_gen5.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48239 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46579 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
```diff
@@ -1569,14 +1569,12 @@ end)
 #
 # In the following two macros, certain identifiers replicate naming conventions
-# defined by C macros in wasm/js/WebAssemblyBuiltin.cpp.
+# defined by C macros in wasm/js/WebAssemblyBuiltin{h, cpp}.
```
Nit put `.` back

kmiller68 commented on Source/JavaScriptCore/runtime/VM.h:
```diff
@@ -639,6 +639,11 @@ class VM : public ThreadSafeRefCountedWithSuppressingSaferCPPChecking<VM>, publi
         return OBJECT_OFFSETOF(VM, m_exception);
     }
 
+    static constexpr ptrdiff_t offsetOfTopCallFrame()
+    {
+        return OBJECT_OFFSETOF(VM, topCallFrame);
+    }
```
We should probably do this via `NativeCallFrameTracer` in the C++ functions for consistency with other operations/slow paths.

ast-hugger replied:
Created https://bugs.webkit.org/show_bug.cgi?id=298109 to do this as a follow-up.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltinTrampoline.cpp:
```diff
@@ -0,0 +1,102 @@
+    jit.loadPtr(JIT::Address(scratch, builtinEntryByteOffset), scratch);
+    jit.storePairPtr(GPRInfo::wasmContextInstancePointer, scratch, GPRInfo::callFrameRegister, JIT::TrustedImm32(CallFrameSlot::codeBlock * sizeof(Register)));
+
+    // Set VM topCallFrame to null to not build an unnecessary stack trace if the function throws an exception.
+    jit.loadPtr(JIT::Address(GPRInfo::wasmContextInstancePointer, JSWebAssemblyInstance::offsetOfVM()), scratch);
```
Doesn't this mean the OOM JS exceptions won't have a stack trace? It seems like we should just use a `NativeCallFrameTracer` so they have a stack. Can you file a bug to fix this?

ast-hugger replied:
Created https://bugs.webkit.org/show_bug.cgi?id=298109

kmiller68 commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -494,17 +495,20 @@ BBQCallee::~BBQCallee()
-{
-    return CodePtr<CFunctionPtrTag>(m_builtin->implementation()).retagged<WasmEntryPtrTag>();
+#if ENABLE(JIT_CAGE)
+    if (Options::useJIT()) {
```
Don't we only need this if we have the JITCage?

ast-hugger replied:
Oops, that's what I meant. :)

kmiller68 reviewed (approved):
r=me with comments

ast-hugger committed:
Enable JIT cage support in Wasm JS String builtins implementation
https://bugs.webkit.org/show_bug.cgi?id=297985
rdar://159302038

Reviewed by Keith Miller.

This patch changes the behavior of WasmBuiltinCallee so that if JIT cage
is enabled, instead of using "static" builtin trampolines defined in
InPlaceInterpreter.asm, we JIT-generate their equivalents. This
ensures that the trampolines are callable from Wasm code. A Wasm builtin
entry point is registered as a JIT operation to make it accessible from
the generated trampoline.

The patch also includes refactorings in WebAssemblyBuiltin.{h, cpp} to:
    - Use more appropriate types for the various code pointers held by
      WebAssemblyBuiltin.
    - Clean up hacky pointer tagging in initializeBuiltinImport()
      in WebAssemblyModuleRecord.cpp.
    - Use more descriptive names for the naming macros (the bulk of
      changes in WebAssemblyBuiltin.cpp).

Canonical link: https://commits.webkit.org/299341@main
